dart/dynamics/InverseKinematics.cpp

```diff
@@ -695,6 +819,420 @@ void InverseKinematics::JacobianTranspose::computeGradient(
+}
+
+//==============================================================================
+void InverseKinematics::Analytical::setQualityComparisonFunction(
+    const QualityComparison& _func)
+{
+  mAnalyticalP.mQualityComparator = _func;
+}
+
+//==============================================================================
+void InverseKinematics::Analytical::resetQualityComparisonFunction()
+{
+  mAnalyticalP.resetQualityComparisonFunction();
+}
+
+//==============================================================================
+InverseKinematics::Analytical::Properties
+InverseKinematics::Analytical::getAnalyticalProperties() const
+{
+  return Properties(getGradientMethodProperties(), mAnalyticalP);
+}
+
+//==============================================================================
+void InverseKinematics::Analytical::constructDofMap()
+{
+  const std::vector<size_t>& analyticalDofs = getDofs();
+  const std::vector<size_t>& nodeDofs =
+      mIK->getNode()->getDependentGenCoordIndices();
+
+  mDofMap.clear();
+  mDofMap.resize(analyticalDofs.size());
+
+  std::vector<bool> isExtraDof;
+  isExtraDof.resize(nodeDofs.size(), true);
+
+  for(size_t i=0; i < analyticalDofs.size(); ++i)
+  {
+    mDofMap[i] = -1;
+    for(size_t j=0; j < nodeDofs.size(); ++j)
+    {
+      if(analyticalDofs[i] == nodeDofs[j])
+      {
+        mDofMap[i] = j;
+        isExtraDof[j] = false;
+      }
+    }
+
+    if(mDofMap[i] == -1)
+    {
+      DegreeOfFreedom* dof = mIK->getNode()->getSkeleton()->
+          getDof(analyticalDofs[i]);
+      std::string name = (dof==nullptr)? std::string("nonexistent") :
+                                         dof->getName();
+      dtwarn << "[InverseKinematics::Analytical::constructDofMap] Your "
+             << "analytical IK solver includes a DegreeOfFreedom ("
+             << analyticalDofs[i] << ") [" << name << "] which is not a "
+             << "dependent DOF of the JacobianNode ["
+             << mIK->getNode()->getName() << "]. This might result in "
+             << "undesirable behavior, such as that DOF being ignored\n";
+    }
+  }
+
+  mExtraDofs.clear();
+  mExtraDofs.reserve(isExtraDof.size());
+
+  const std::vector<int>& gradDofMap = mIK->getDofMap();
+  for(size_t i=0; i < isExtraDof.size(); ++i)
+  {
+    if( isExtraDof[i] && (gradDofMap[i] > -1) )
+      mExtraDofs.push_back(i);
+  }
+}
+
+//==============================================================================
+void InverseKinematics::Analytical::checkSolutionJointLimits()
+{
+  const std::vector<size_t>& dofs = getDofs();
+  for(size_t i=0; i < mSolutions.size(); ++i)
+  {
+    Solution& s = mSolutions[i];
+    const Eigen::VectorXd& q = s.mConfig;
+
+    for(size_t j=0; j < dofs.size(); ++j)
+    {
+      DegreeOfFreedom* dof = mIK->getNode()->getSkeleton()->getDof(dofs[j]);
+      if(q[j] < dof->getPositionLowerLimit()
+         || dof->getPositionUpperLimit() < q[j])
+      {
+        s.mValidity |= LIMIT_VIOLATED;
```

Comment:
I agree with that the user might not want DART to overwrite it in some cases. We can consider to pass a flag that controls the behavior, and let the user to select the behavior. 

I don't have strong preference here so I would be fine with it as it is now, but, for clarification, I would like to make a note about the behavior of whichever we take.